src/adc.rs

```diff
@@ -0,0 +1,268 @@
+    pub fn disable(&mut self) {
+        let adc = unsafe{ &*stm32::ADC::ptr() };
+        adc.ccr.write(|w| w.vrefen().clear_bit());
+    }
+}
+
+impl VBat {
+    pub fn new() -> Self { VBat {} }
+
+    /// Enable the internal VBat sense, remember to disable when not in use
+    /// as otherwise it will sap current from the VBat source.
+    pub fn enable(&mut self) {
+        let adc = unsafe{ &*stm32::ADC::ptr() };
+        adc.ccr.write(|w| w.vbaten().set_bit());
+    }
+
+    /// Disable the internal VBat sense.
+    pub fn disable(&mut self) {
+        let adc = unsafe{ &*stm32::ADC::ptr() };
+        adc.ccr.write(|w| w.vbaten().clear_bit());
+    }
+}
+
+impl Adc {
+    pub fn new(adc: stm32::ADC) -> Self {
+        let mut s = Self {
+            rb: adc,
+            sample_time: AdcSampleTime::default(),
+        };
+        s.select_clock();
+        s.calibrate();
+        s
+    }
+
+    pub fn set_sample_time(&mut self, t_samp: AdcSampleTime) {
+        self.sample_time = t_samp;
+    }
+
+    fn calibrate(&mut self) {
+        /* Ensure that ADEN = 0 */
+        if self.rb.cr.read().aden().bit_is_set() {
+            /* Clear ADEN by setting ADDIS */
+            self.rb.cr.write(|w| w.addis().set_bit());
+        }
+        while self.rb.cr.read().aden().bit_is_set() {}
+
+        /* Clear DMAEN */
+        self.rb.cfgr1.write(|w| w.dmaen().clear_bit());
+
+        /* Start calibration by setting ADCAL */
+        self.rb.cr.write(|w| w.adcal().set_bit());
+
+        /* Wait until calibration is finished and ADCAL = 0 */
+        while self.rb.cr.read().adcal().bit_is_set() {}
+    }
+
+    fn select_clock(&mut self) {
+        let rcc = unsafe { &*stm32::RCC::ptr()};
+
+        rcc.apb2enr.write(|w| w.adcen().set_bit());
+        rcc.cr2.write(|w| w.hsi14on().set_bit());
+        while rcc.cr2.read().hsi14rdy().bit_is_clear() {}
+    }
+
+    fn power_up(&mut self) {
+        if self.rb.isr.read().adrdy().bit_is_set() {
+            self.rb.isr.write(|w| w.adrdy().clear_bit());
+        }
+        self.rb.cr.write(|w| w.aden().set_bit());
+        while self.rb.isr.read().adrdy().bit_is_clear() {}
+    }
+
+    fn power_down(&mut self) {
+        self.rb.cr.write(|w| w.adstp().set_bit());
+        while self.rb.cr.read().adstp().bit_is_set() {}
+        self.rb.cr.write(|w| w.addis().set_bit());
+        while self.rb.cr.read().aden().bit_is_set() {}
+    }
+
+    fn convert(&mut self, chan: u8) -> u16 {
+        self.rb.chselr.write(|w| unsafe { w.bits(1_u32 << chan) });
+        self.sample_time.write_bits(&mut self.rb);
+        self.rb.cr.write(|w| w.adstart().set_bit());
+        while self.rb.isr.read().eoc().bit_is_clear() {}
+        self.rb.dr.read().bits() as u16
+    }
+}
+
+impl<WORD, PIN> OneShot<Adc, WORD, PIN> for Adc
+    where
+        WORD: From<u16>,
+        PIN: Channel<Adc, ID=u8>,
+{
+    type Error = ();
+
+    fn read(&mut self, _pin: &mut PIN) -> nb::Result<WORD, Self::Error> {
```

Comment:
I'd at least add a comment that not all bits are used. 
I'd prefer just shifting it and leaving the lower bits always at zero, seems like a less surprising interface to me.

Reply:
The default setup does this.

Reply:
As it is setup by default the 12 bits get stored in the lower twelve bits of the data register, essentially producing a u16 with a value of 0-4095. The lower precision modes also shift right as appropriate so they also present as 0-n values.

Reply:
No, i mean exactly the opposite. Few people like reading docs and per default I'd expect the u16 value to go from 0 to 0xFFFF (technically 0xFFF0 in this case). This also makes the trait non-portable, since  potential libraries would probably expect a normal u16 value.

Reply:
Ah, so we would lose the least 4 significant bits. That is definitely do-able, though introduces a weird quirk at 6bit res.

Reply:
Possibly I could include an optional mode if you want instead numbers always from 0-N. I was planning anyway on adding something that displayed the max range value.